File: app/src/main/java/com/imangazalievm/bubbble/data/repository/ShotsRepository.java
package com.imangazalievm.bubbble.data.repository;

import com.imangazalievm.bubbble.data.repository.datastores.ShotsDataStore;
import com.imangazalievm.bubbble.domain.models.Shot;
import com.imangazalievm.bubbble.domain.models.ShotsRequestParams;
import com.imangazalievm.bubbble.domain.models.UserShotsRequestParams;
import com.imangazalievm.bubbble.domain.repository.IShotsRepository;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;

public class ShotsRepository implements IShotsRepository {

    private ShotsDataStore shotsDataStore;

    @Inject
    public ShotsRepository(ShotsDataStore shotsDataStore) {
        this.shotsDataStore = shotsDataStore;
    }

    @Override
    public Single<List<Shot>> getShot(ShotsRequestParams shotsRequestParams) {
        return shotsDataStore.getShots(shotsRequestParams);
    }

    @Override
    public Single<Shot> getShot(long shotId) {
        return shotsDataStore.getShot(shotId);
    }

    @Override
    public Single<List<Shot>> getUserShots(UserShotsRequestParams requestParams) {
        return shotsDataStore.getUserShots(requestParams);
    }

}

File: core/data/src/main/java/com/bubbble/data/shots/ShotsRepository.kt
package com.bubbble.data.shots

import androidx.paging.Pager
import androidx.paging.PagingConfig
import com.bubbble.core.models.shot.Shot
import com.bubbble.core.models.feed.ShotsFeedParams
import com.bubbble.core.models.search.SearchParams
import com.bubbble.core.models.shot.ShotDetails
import com.bubbble.core.models.shot.ShotDetailsParams
import com.bubbble.core.models.user.UserShotsParams
import com.bubbble.core.network.DribbbleApi
import com.bubbble.data.global.parsing.PageParserManager
import com.bubbble.data.shots.feed.FeedShotsParser
import com.bubbble.data.global.paging.CommonPagingSource
import com.bubbble.data.shots.details.ShotDetailsParser
import com.bubbble.data.shots.search.SearchPageParser
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ShotsRepository @Inject constructor(
    private val dribbbleApi: DribbbleApi,
    private val pageParserManager: PageParserManager,
    private val feedShotsParser: FeedShotsParser,
    private val searchPageParser: SearchPageParser,
    private val shotDetailsParser: ShotDetailsParser
) {

    fun getShots(
        requestParams: ShotsFeedParams
    ) = Pager(PagingConfig(pageSize = 20)) {
        CommonPagingSource { pagingParams ->
            pageParserManager.parse(
                feedShotsParser,
                requestParams,
                pagingParams
            )
        }
    }.flow

    suspend fun getShot(shotSlug: String): ShotDetails {
        return pageParserManager.parse(shotDetailsParser, ShotDetailsParams(shotSlug))
    }

    suspend fun getUserShots(requestParams: UserShotsParams): List<Shot> {
        TODO()
        //return dribbbleApi.getUserShots(
        //    requestParams.userName,
        //    requestParams.page,
        //    requestParams.pageSize
        //)
    }

    suspend fun search(
        params: SearchParams
    ) = Pager(PagingConfig(pageSize = 20)) {
        CommonPagingSource { pagingParams ->
            pageParserManager.parse(searchPageParser, params, pagingParams)
        }
    }.flow

}